proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -219,24 +251,160 @@ lemma reply_unlink_tcb_corres:
   (wp: crunch_wps simp: crunch_simps)
 
+lemma set_endpoint_valid_tcb[wp]:
+  "set_endpoint ep endpoint \<lbrace>valid_tcb ptr tcb\<rbrace>"
+  apply (clarsimp simp: set_simple_ko_def)
+  apply (wpsimp wp: set_object_typ_ats get_object_wp)+
+  done
+
+lemma set_endpoint_valid_tcbs[wp]:
+  "set_endpoint ep endpoint \<lbrace>valid_tcbs\<rbrace>"
+  unfolding valid_tcbs_def
+  apply (wpsimp wp: hoare_vcg_ball_lift2 hoare_vcg_all_lift)
+    apply (wpsimp wp: set_simple_ko_wps)
+    apply (clarsimp simp: obj_at_def is_ep_def)
+   apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift')
+     apply (wpsimp wp: set_simple_ko_wps)
+    apply wpsimp
+   apply (clarsimp simp: obj_at_def)+
+  done
+
+lemma set_notification_valid_tcb[wp]:
+  "set_notification ntfnptr val \<lbrace>valid_tcb ptr tcb\<rbrace>"
+  apply (clarsimp simp: set_simple_ko_def)
+  apply (wpsimp wp: get_object_wp)
+  done
+
+lemma set_notification_valid_tcbs[wp]:
+  "set_notification ntfnptr val \<lbrace>valid_tcbs\<rbrace>"
+  unfolding valid_tcbs_def
+  apply (wpsimp wp: hoare_vcg_ball_lift2)
+    apply (wpsimp wp: set_simple_ko_wp)
+    apply (clarsimp simp: simple_obj_at_def obj_at_def)
+   apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift')
+     apply (wpsimp wp: set_simple_ko_wp)
+    apply wpsimp
+   apply (clarsimp simp: simple_obj_at_def obj_at_def)+
+  done
+
+lemma setNotification_valid_tcb'[wp]:
+  "setNotification ntfn val \<lbrace>valid_tcb' tcb\<rbrace>"
+  apply (clarsimp simp: setNotification_def)
+  apply (rule setObject_valid_tcb')
+  done
+
+lemma setNotification_valid_tcbs'[wp]:
+  "setNotification ntfn val \<lbrace>valid_tcbs'\<rbrace>"
+  unfolding valid_tcbs'_def
+  by (wpsimp wp: set_ntfn'.setObject_wp hoare_vcg_all_lift hoare_vcg_imp_lift'
+           simp: setNotification_def)+
+
+lemma setEndpoint_valid_tcb'[wp]:
+  "setEndpoint epPtr val \<lbrace>valid_tcb' tcb\<rbrace>"
+  apply (clarsimp simp: setEndpoint_def)
+  apply (rule setObject_valid_tcb')
+  done
+
+lemma setEndpoint_valid_tcbs'[wp]:
+  "setEndpoint ePtr val \<lbrace>valid_tcbs'\<rbrace>"
+  unfolding valid_tcbs'_def
+  by (wpsimp wp: set_ep'.setObject_wp hoare_vcg_all_lift hoare_vcg_imp_lift'
+           simp: setEndpoint_def)+
+
+(* FIXME RT MICHAEL: move to where invs_queues is *)
+lemma invs_queues'[elim!]:
+  "invs' s \<Longrightarrow> valid_queues' s"
+  by (simp add: invs'_def valid_state'_def)
+
+lemma update_valid_tcb:
+  "valid_tcb ptr tcb s \<Longrightarrow> valid_tcb ptr tcb (s\<lparr>ready_queues := a\<rparr>)"
+  "valid_tcb ptr tcb s \<Longrightarrow> valid_tcb ptr tcb (s\<lparr>scheduler_action := b\<rparr>)"
+  apply (clarsimp simp: valid_tcb_def valid_tcb_state_def obj_at_def split: thread_state.splits)
+  apply (intro conjI impI)
+  apply (case_tac "tcb_state tcb"; clarsimp?)
+  apply (clarsimp simp: obj_at_def is_reply_def split: option.splits)
+  apply (clarsimp simp: valid_bound_obj_def split: option.splits)+
+  apply (clarsimp simp: ARM.valid_arch_tcb_def)+
```

Comment:
I couldn't find much about valid_arch_tcb. I'm guessing it's not OK to have the ARM. definition in here. What's the approach here?

Reply:
This is in an `ARM` subdirectory, so it'd be Ok. The question is more why we're not inside the `Arch` locale here and therefore just `valid_arch_tcb_def` doesn't work.

Reply:
You might be able to wrap the whole thing in an 

```
context begin interpretation Arch .
...
end
```

Reply:
Wrapping it in context worked, thanks!